src/CommandLine.cc

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (C) 1996-2018 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+
+#include "CommandLine.h"
+
+static void
+ResetGetopt(const bool allowStderrWarnings)
+{
+    opterr = allowStderrWarnings ? 1 : 0;
+    // getopt(3) uses global state but resets it if optind is zero
+    // getopt(3) always skips argv[0], even if optind is zero
+    optind = 0;
+}
+
+CommandLine::CommandLine(int argC, char *argV[], const char *shortRules, const struct option *longRules):
```

Comment:
Note that if shortRules is null, getopt_long() aborts with SIGSEGV. I think we need an assertion.

Reply:
I suggest either using a reference (instead of a pointer) or adding an internal getopt() call (instead of getopt_long(), only when there are no long options).

Reply:
Done with the reference.